gateway/mw_jwt.go

```diff
@@ -52,13 +68,67 @@ func parseJWK(buf []byte) (*jose.JSONWebKeySet, error) {
 	return &j, nil
 }
 
+func (k *JWTMiddleware) legacyGetSecretFromURL(url, kid, keyType string) (interface{}, error) {
+	// Implement a cache
+	if JWKCache == nil {
+		JWKCache = cache.New(240*time.Second, 30*time.Second)
+	}
+
+	var jwkSet JWKs
+	cachedJWK, found := JWKCache.Get("legacy-" + k.Spec.APIID)
+	if !found {
+		resp, err := http.Get(url)
+		if err != nil {
+			k.Logger().WithError(err).Error("Failed to get resource URL")
+			return nil, err
+		}
+		defer resp.Body.Close()
+
+		// Decode it
+		if err := json.NewDecoder(resp.Body).Decode(&jwkSet); err != nil {
+			k.Logger().WithError(err).Error("Failed to decode body JWK")
+			return nil, err
+		}
+
+		JWKCache.Set("legacy-" + k.Spec.APIID, jwkSet, cache.DefaultExpiration)
+	} else {
+		jwkSet = cachedJWK.(JWKs)
+	}
+
+	for _, val := range jwkSet.Keys {
+		if val.KID != kid || strings.ToLower(val.Kty) != strings.ToLower(keyType) {
+			continue
+		}
+		if len(val.X5c) > 0 {
+			// Use the first cert only
+			decodedCert, err := base64.StdEncoding.DecodeString(val.X5c[0])
+			if !bytes.Contains(decodedCert, []byte("-----")) {
+				return nil, errors.New("No legacy public keys found")
+			}
+			if err != nil {
+				return nil, err
+			}
+			return ParseRSAPublicKey(decodedCert)
+		}
+		return nil, errors.New("no certificates in JWK")
+	}
+
+	return nil, errors.New("No matching KID could be found")
+}
+
 func (k *JWTMiddleware) getSecretFromURL(url, kid, keyType string) (interface{}, error) {
 	// Implement a cache
 	if JWKCache == nil {
 		k.Logger().Debug("Creating JWK Cache")
 		JWKCache = cache.New(240*time.Second, 30*time.Second)
 	}
 
+	if key, err := k.legacyGetSecretFromURL(url, kid, keyType); err == nil {
+		return key, nil
+	} else {
+		k.Logger().WithError(err).Info("Legacy JWK qery")
+	}
```

Comment:
I think this should be a fallback. When we fail to get certificates in the standard way. `parseJWK` will fail if we have keys inside `x5c` , I think this should happen after `parseJWK` returns an error.

 Also, we don't need to make 2 calls to the JWK endpoint just to receive the same data( when legacy fails, we have to try again with standard). We can call the endpoint once, read the body into a buffer and use the buffer for both standard and legacy mode.

Reply:
Yes I thought about it. It required bigger rewrite. But at the same time we do caching so performance not really an issue. Any ideas on it without rewriting too much stuff?

Also note that once cached it will fail fast since just check if blob has prefix.

Reply:
@buger 

something like this will do

```
		if jwkSet, err = parseJWK(buf); err != nil {
			k.Logger().WithError(err).Error("Failed to decode standard body JWK")
			k.Logger().Debug("Trying legacy JWK")
			return k.legacyGetSecretFromURL(url, kid, keyType)
		}
```

Reply:
`legacyGetSecretFromURL` can accept `buf`  and this will avoid additional call with minimal refactoring.

Reply:
Done